### PR TITLE
test: Smoke Tests for packaged embeds and build improvements

### LINKED ISSUE #9091: [CAL-1757] [CI Automation]Make an independent react application that uses @calcom/embed-react and verify that it builds fine.
This is needed because a built package when published to npm might have certain files missing and types are imported from *.d.ts files instead of .ts files directly in that case.

<sub>[CAL-1757](https://linear.app/calcom/issue/CAL-1757/[ci-automation]make-an-independent-react-application-that-uses)</sub>

### LINKED ISSUE #9168: [CAL-1805] Embed: The current build process has no checks to prevent sending wrong embed URL in built code.
It happened recently that embed-snippet was built without these explicity env variables.`NEXT_PUBLIC_EMBED_LIB_URL='https://app.cal.com/embed/embed.js' NEXT_PUBLIC_WEBAPP_URL='https://app.cal.com'`

When I published embed-react it used that build which had localhost:3000 references in it.

Ensure that when `publish-embed` command is used, all embed packages are rebuild with correct env variables.

Originally reported here: [https://calendso.slack.com/archives/C01NXDSRY9W/p1685341062862069?thread_ts=1685310906.817079&cid=C01NXDSRY9W](https://calendso.slack.com/archives/C01NXDSRY9W/p1685341062862069?thread_ts=1685310906.817079&cid=C01NXDSRY9W)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-1805](https://linear.app/calcom/issue/CAL-1805/embed-the-current-build-process-has-no-checks-to-prevent-sending-wrong)</sub>

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

Adds tests for the packaged code and improved package publlishing flow to avoid 2 critical issues reported recently by customers. 
- TypeScript types issue couldn't be caught in e2e because we don't import @calcom/embed-react there
- Packaged code with http://localhost:3000 couldn't be caught because when I tested locally or in CI that URL is available but in production obviously that won't be available. For now that's avoided by ensuring that during npm publish the packages are rebuilt(with existing dist being cleaned-up) for sure with correct env variables.

Fixes #9168 Fixes CAL-1805
Fixes #9091 Fixes CAL-1757

## Type of change
- Tests

## How should this be tested?

- [x] Run packaged code tests with  `yarn test -- --packaged-embed-tests-only`. It would need you to build` embed-react ` first using `cd packages/embeds/embed-react && yarn build`
- [x] Run `yarn publish-embed` command after deleting `dist` from all embed packages(embed-core, embed-react,embed-snippet). It would generate `dist` folder automatically with correct URL https://app.cal.com

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,7 +18,8 @@
     "app-store": "yarn app-store-cli cli",
     "create-app": "yarn app-store create",
     "edit-app": "yarn app-store edit",
-    "publish-embed": "yarn workspaces foreach --from=\"@calcom/embed*\" npm publish --access=public",
+    "withHostedCalcomEnv": "NEXT_PUBLIC_EMBED_LIB_URL='https://app.cal.com/embed/embed.js' NEXT_PUBLIC_WEBAPP_URL='https://app.cal.com' yarn",
+    "publish-embed": "yarn withHostedCalcomEnv workspace @calcom/embed-core build && yarn withHostedCalcomEnv workspace @calcom/embed-snippet build && yarn workspaces foreach --from=\"@calcom/embed*\" npm publish --access=public",
```
Comment (hariombalhara) [inline package.json:22]: embed-core and embed-snippet are the dependencies for embed-react and any new framework specific package(that might be added). So, build them first.

File: packages/embeds/embed-core/package.json [hunk 2]
```diff
@@ -15,7 +15,7 @@
     "embed-web-start": "yarn workspace @calcom/web start",
     "__build": "yarn tailwind && vite build && tsc --emitDeclarationOnly --declarationDir dist && cp -r ../../../apps/web/public/embed ./dist/",
     "__dev": "yarn tailwind && vite build --mode development",
-    "build": "NEXT_PUBLIC_EMBED_FINGER_PRINT=$(git rev-parse --short HEAD) yarn __build",
+    "build": "rm -rf dist && NEXT_PUBLIC_EMBED_FINGER_PRINT=$(git rev-parse --short HEAD) yarn __build",
```
Comment (hariombalhara) [inline packages/embeds/embed-core/package.json:18]: Ensure that build is cleaned before doing a fresh build. it ensures that stale files/changes aren't pushed to npm.

File: .github/workflows/e2e-embed-react.yml [hunk 3]
```diff
@@ -26,6 +26,7 @@ jobs:
       - uses: ./.github/actions/cache-build
       - name: Run Tests
         run: yarn test-e2e:embed-react
+        run: yarn workspace @calcom/embed-react packaged:tests
```
Comment (hariombalhara) [inline .github/workflows/e2e-embed-react.yml:29]: Runs packaged code tests.

File: package.json [hunk 4]
```diff
@@ -17,7 +17,8 @@
     "app-store": "yarn app-store-cli cli",
     "create-app": "yarn app-store create",
     "edit-app": "yarn app-store edit",
-    "publish-embed": "yarn workspaces foreach --from=\"@calcom/embed*\" npm publish --access=public",
+    "withEmbedPublishEnv": "NEXT_PUBLIC_EMBED_LIB_URL='https://app.cal.com/embed/embed.js' NEXT_PUBLIC_WEBAPP_URL='https://app.cal.com' yarn",
+    "publish-embed": "yarn withEmbedPublishEnv workspace @calcom/embed-core build && yarn withEmbedPublishEnv workspace @calcom/embed-snippet build && yarn workspaces foreach --from=\"@calcom/embed*\" npm publish --access=public",
```
Comment (hariombalhara) [inline package.json:21]: Build embed-core and embed-snippet first which are required by embed-react and any new framework embed that we add.

File: packages/embeds/embed-core/playwright/lib/testUtils.ts [hunk 5]
```diff
@@ -131,9 +131,9 @@ export async function bookFirstEvent(username: string, frame: Frame, page: Page)
   return booking;
 }
 
-export async function rescheduleEvent(username, frame, page) {
+export async function rescheduleEvent(username: string, frame: Frame, page: Page) {
```
Comment (hariombalhara) [inline packages/embeds/embed-core/playwright/lib/testUtils.ts:134]: Type improvements.

File: packages/embeds/embed-react/package.json [hunk 6]
```diff
@@ -11,23 +11,24 @@
...
     "preview": "vite preview",
-    "prepare": "NEXT_PUBLIC_EMBED_LIB_URL='https://app.cal.com/embed/embed.js' NEXT_PUBLIC_WEBAPP_URL='https://app.cal.com'  yarn build",
     "type-check": "tsc --pretty --noEmit",
     "type-check:ci": "tsc-absolute --pretty --noEmit",
     "lint": "eslint --ext .ts,.js,.tsx,.jsx ./src",
     "embed-tests": "yarn playwright test --config=./playwright/config/playwright.config.ts",
     "embed-tests-quick": "QUICK=true yarn embed-tests",
     "embed-tests-update-snapshots:ci": "yarn embed-tests-quick --update-snapshots",
-    "prepack": "NEXT_PUBLIC_EMBED_LIB_URL='https://app.cal.com/embed/embed.js' NEXT_PUBLIC_WEBAPP_URL='https://app.cal.com' yarn build",
+    "packaged:tests": "cd test/packaged && yarn tsc --noEmit && yarn run -T test -- --packaged-embed-tests-only",
```
Comment (hariombalhara) [inline packages/embeds/embed-react/package.json:22]: run test command from root but with packaged-embed-tests only.

File: packages/embeds/embed-react/package.json [hunk 7]
```diff
@@ -11,23 +11,24 @@
...
+    "withEmbedPublishEnv": "NEXT_PUBLIC_EMBED_LIB_URL='https://app.cal.com/embed/embed.js' NEXT_PUBLIC_WEBAPP_URL='https://app.cal.com' yarn",
+    "prepack": "yarn withEmbedPublishEnv build && yarn packaged:tests",
     "embed-web-start": "yarn workspace @calcom/web start",
     "embed-dev": "yarn workspace @calcom/embed-react dev",
     "clean": "rm -rf .turbo && rm -rf node_modules && rm -rf dist"
   },
   "main": "./dist/Cal.umd.js",
   "module": "./dist/Cal.es.js",
-  "types": "./dist/index.d.ts",
+  "types": "./dist/embed-react/src/index.d.ts",
```
Comment (hariombalhara) [inline packages/embeds/embed-react/package.json:31]: This is actually fixed already in https://github.com/calcom/cal.com/pull/9090 which is yet to be merged.

File: packages/embeds/embed-react/src/Cal.tsx [hunk 8]
```diff
@@ -11,7 +13,7 @@ type CalProps = {
     debug?: boolean;
     uiDebug?: boolean;
   };
-  config?: Record<string, string>;
+  config?: PrefillAndIframeAttrsConfig;
```
Comment (hariombalhara) [inline packages/embeds/embed-react/src/Cal.tsx:16]: Improves autocompletion for @calcom/embed-react users.

File: vitest.workspace.ts [hunk 9]
```diff
@@ -1,20 +1,32 @@
...
+      {
+        test: {
+          include: ["packages/embeds/**/*.{test,spec}.{ts,js}"],
+          environment: "jsdom",
+        },
+      },
+    ]
+  : [
+      {
+        test: {
```
Comment (hariombalhara) [inline vitest.workspace.ts:16]: Don't slowdown core tests for packaged code tests(which require build for all embed packages).

File: packages/embeds/embed-react/tsconfig.json [hunk 10]
```diff
@@ -13,6 +13,7 @@
       "@calcom/embed-snippet": ["../embed-snippet/src"]
     }
   },
-  "include": ["src", "env.d.ts"],
-  "exclude": ["dist", "build", "node_modules", "test-cal.tsx"]
+  "include": ["**/*.ts","**/*.tsx", "env.d.ts"],
```
Comment (hariombalhara) [inline packages/embeds/embed-react/tsconfig.json:16]: Validate all .ts and .tsx files now. Some files like inline.tsx were outside src that are used as playground but they can be good for verifying the EMbed API types as well.

File: .github/workflows/e2e-embed-react.yml (hunk 3 above)
Comment (zomars) [inline .github/workflows/e2e-embed-react.yml:29]: This is invalid syntax

```suggestion
        run: |
          yarn test-e2e:embed-react
          yarn workspace @calcom/embed-react packaged:tests
```

**Review by zomars — DISMISSED** [review]
Love the idea of having these tests. And it seems they're working! We got out first error:

image

Also left some syntax fixes to address before merging.

Nice work so far @hariombalhara :pray:

File: .github/workflows/e2e-embed-react.yml (hunk 3 above)
Comment (hariombalhara) [reply .github/workflows/e2e-embed-react.yml:29]: Is it invalid syntax or not a recommended approach? Syntax wise it's working already.

File: .github/workflows/e2e-embed-react.yml [hunk 11]
```diff
@@ -25,7 +25,9 @@ jobs:
       - uses: ./.github/actions/cache-db
       - uses: ./.github/actions/cache-build
       - name: Run Tests
-        run: yarn test-e2e:embed-react
+        run: |
+          yarn test-e2e:embed-react
+          yarn workspace @calcom/embed-react packaged:tests
```
Comment (hariombalhara) [inline .github/workflows/e2e-embed-react.yml:30]: Results for the new test.
https://github.com/calcom/cal.com/actions/runs/5130425921

**Review by zomars — APPROVED** [review]
LGTM 

Just waiting for checks to pass 🙏🏽

**Review by zomars — APPROVED** [review]
LGTM 

Just waiting for checks to pass 🙏🏽

**Pull request merged**